[FLINK-13290][table-planner-blink][hbase] SinkCodeGenerator should not compare row type field names and enable blink planner for hbase IT case

wuchong opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Fix `SinkCodeGenerator` shouldn't compare type field names. Also we added blink planner for hbase IT cases.

## Brief change log

- `SinkCodeGenerator` still use type information to compare
- make hbase IT case to run with blink planner and flink planner both.

## Verifying this change

- We added some IT case in blink planner to verify `SinkCodeGenerator`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

wuchong commented:
cc @carp84  could you help review the hbase part?
cc @JingsongLi @godfreyhe welcome to review when you are free.

JingsongLi commented on flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/HBaseLookupSourceITCase.java:
```diff
@@ -163,6 +165,7 @@ private static Put putRow(int rowKey, int f1c1, String f2c1, long f2c2, double f
 		return descriptorProperties.asMap();
 	}
 
+	@Ignore
```
Why ignore?

JingsongLi commented on flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/HBaseSinkITCase.java:
```diff
@@ -52,9 +55,13 @@
 public class HBaseSinkITCase extends HBaseTestingClusterAutostarter {
 	private static final long serialVersionUID = 1L;
 
+	private static final String BLINK_PLANNER = "Blink";
```
use enum instead?

JingsongLi commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/SinkCodeGenerator.scala:
```diff
@@ -108,8 +108,7 @@ object SinkCodeGenerator {
     val fieldIndexProcessCode =
-      if (getCompositeTypes(convertOutputType).map(fromTypeInfoToLogicalType) sameElements
-          inputTypeInfo.getFieldTypes.map(fromTypeInfoToLogicalType)) {
+      if (getCompositeTypes(convertOutputType) sameElements inputTypeInfo.getFieldTypes) {
```
it not work now...
`inputTypeInfo` is a BaseRowTypeInfo with internal type informations.
And `convertOutputType` is the type information from sink and users.
Consider Types.STRING and BinaryStringTypeInfo, Types.DECIMAL and DecimalTypeInfo.

JingsongLi replied:
And I think maybe you can try `LogicalType.asSerializableString`

carp84 commented on flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/HBaseLookupSourceITCase.java:
```diff
@@ -200,4 +203,44 @@ public void testHBaseLookupFunction() throws Exception {
+	public void testHBaseLookupTableSource() throws Exception {
+		EnvironmentSettings settings = EnvironmentSettings.newInstance()
+			// only blink planner support lookup table source
+			.useBlinkPlanner().inStreamingMode().build();
```
We should use parameterized test to both cover blink-planner and flink-planner just like the way in `HBaseSinkITCase`, and remove the `testHBaseLookupFunction` test.

wuchong commented:
Hi @carp84 , @JingsongLi , I have updated the PR.

I split it into 3 commits:

- 0a59bfc570f793a54b2b3d14063858934a67a1fc: Add method to equals LogicalType without field names
- 66492281e2ad34ac737ae7f87945ab60276ea0ed: SinkCodeGenerator should not compare row type field names
- cef04c07f422c8d0190780e6dc3ab5a02dc798a2: Enable blink planner for integration tests of HBase


cc @twalthr , it would be great if you can help to review 0a59bfc570f793a54b2b3d14063858934a67a1fc.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeChecks.java:
```diff
@@ -145,6 +156,14 @@ public static boolean isSingleFieldInterval(LogicalType logicalType) {
 	}
 
+	public static boolean areTypeEqualsWithoutNames(LogicalType thisType, LogicalType thatType) {
+		if (thisType == null || thatType == null) {
```
It's better to add a precondition here. Types should not be null. If they are null, it is a bug that should throw an exception.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeChecks.java:
```diff
@@ -332,4 +351,162 @@ public Boolean visit(DayTimeIntervalType dayTimeIntervalType) {
 		}
 	}
+
+	private static class TypeEquivalenceIgnoreNamesExtractor extends Extractor<Boolean> {
```
Should directly extend from `LogicalTypeDefaultVisitor` because we override the default method anyway.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeChecks.java:
```diff
@@ -332,4 +351,162 @@ public Boolean visit(DayTimeIntervalType dayTimeIntervalType) {
+		 * Returns true if tow LogicalType can equal. This is the same with {@link LogicalType#equals(Object)}.
+		 */
+		private boolean canEqual(LogicalType thatType) {
+			if (thisType == null || thatType == null) {
```
same comment as above regarding `null` checking

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeChecks.java:
```diff
@@ -332,4 +351,162 @@ public Boolean visit(DayTimeIntervalType dayTimeIntervalType) {
+			if (!canEqual(thatType)) {
+				return false;
+			}
+			boolean thisIsFinal = ((DistinctType) thisType).isFinal();
```
Distinct types are always final. We don't need to check that.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeChecks.java:
```diff
@@ -145,6 +156,14 @@ public static boolean isSingleFieldInterval(LogicalType logicalType) {
 		return logicalType.accept(SINGLE_FIELD_INTERVAL_EXTRACTOR);
 	}
 
+	public static boolean areTypeEqualsWithoutNames(LogicalType thisType, LogicalType thatType) {
```
nit: rename to `hasEqualTypesWithoutNames`

wuchong replied:
What's the reason to use `hasEqualTypesWithoutNames`? It looks like whether it contains some equal types (might children types are equal).

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeChecks.java:
```diff
@@ -332,4 +351,162 @@ public Boolean visit(DayTimeIntervalType dayTimeIntervalType) {
+				RowType.RowField thatField = thatFields.get(i);
+				// ignore field names
+				if (!thisField.getType().equals(thatField.getType())
+					|| !Objects.equals(thisField.getDescription(), thatField.getDescription())) {
```
We do we check the description? Isn't it as meaningless as the field names?

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeChecks.java:
```diff
@@ -332,4 +351,162 @@ public Boolean visit(DayTimeIntervalType dayTimeIntervalType) {
+			return thisIsFinal == thatType.isFinal() &&
+				thisObjectIdentifier.equals(thatType.getObjectIdentifier()) &&
+				thisDescription.equals(thatType.getDescription()) &&
+				areTypeEqualsWithoutNames(thisSourceType, thatType.getSourceType());
```
I think we can just compare the source type. Everything else is logical meta data, no?

wuchong commented:
Hi @twalthr , thanks for the advice. The implementation of type equals has been much simpler.

Please have a look again.

carp84 commented on flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/util/HBaseTestBase.java:
```diff
@@ -0,0 +1,157 @@
+		};
+		// split keys
+		byte[][] splitKeys = new byte[][]{ Bytes.toBytes(4) };
+		createTable(tableName, families, splitKeys);
```
Minor: we could extract L101-L111 out to remove some duplicated codes.

carp84 commented on flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/util/HBaseTestingClusterAutostarter.java:
```diff
@@ -74,7 +75,7 @@
 // https://github.com/apache/hbase/blob/master/hbase-server/src/test/java/org/apache/hadoop/hbase/filter/FilterTestingCluster.java
 //
-public class HBaseTestingClusterAutostarter extends TestLogger implements Serializable {
+public abstract class HBaseTestingClusterAutostarter extends AbstractTestBase {
```
Minor: HBaseTestingClusterAutostarter -> HBaseTestingClusterAutoStarter

carp84 commented on flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/HBaseConnectorITCase.java:
```diff
@@ -20,281 +18,464 @@
+	@Test
+	public void testTableInputFormat() throws Exception {
+		if (BLINK_PLANNER.equals(planner)) {
+			// this case is for testing TableInputFormat which is not works for flink-table
```
This comment is a little bit ambiguous, from the code logic it seems we should say "this case is for testing TableInputFormat which only applies to flink-table planner" or so?

wuchong replied:
Not exactly. I mean `TableInputFormat` is a runtime implementation, not for flink table, not relative to what the planner is. You can see the test is only verified on `DataSet`, not on `TableEnvironment`.

carp84 replied:
After some offline discussion, the case is irrelative to planner actually, but in case that later change somehow making the test case relative to planner implementation (although low possibility), let's just don't skip the test for any planner.

carp84 commented on flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/util/HBaseTestBase.java:
```diff
@@ -0,0 +1,157 @@
+
+package org.apache.flink.addons.hbase.util;
+
+import org.apache.flink.addons.hbase.HBaseConnectorITCase2;
```
Oops, it seems this issue is already resolved by the later force push, and now the code here looks good.

wuchong commented:
As discussed with @twalthr , we would like to go with `areTypeCompatible` for the method. And add a detailed javadoc on the method.

PR updated and squashed.
